File: src/main/java/pl/dmichalski/algorithms/_9_sorting/insertion/InsertionSortService.java
package pl.dmichalski.algorithms._9_sorting.insertion;

class InsertionSortService {

    /**
     * O(n2) time complexity
     * O(1) space complexity
     */
    int[] sort(int[] values) {
        if (values == null) {
            return null;
        }

        for (int i = 1; i < values.length; i++) {
            int currentVal = values[i];
            int j;
            for (j = i - 1; j >= 0 && values[j] > currentVal; j--) {
                values[j + 1] = values[j];
            }
            values[j + 1] = currentVal;
        }

        return values;
    }

}

File: src/main/kotlin/pl/dmichalski/algorithms/_9_sorting/insertion/InsertionSortService.kt
package pl.dmichalski.algorithms._9_sorting.insertion

import pl.dmichalski.algorithms._9_sorting.SortService

internal class InsertionSortService : SortService {

    /**
     * O(n2) time complexity
     * O(1) space complexity
     */
    override fun sort(values: IntArray): IntArray {
        for (i in 1 until values.size) {
            val currentVal = values[i]
            var j: Int = i - 1
            while (j >= 0 && values[j] > currentVal) {
                values[j + 1] = values[j]
                j--
            }
            values[j + 1] = currentVal
        }
        return values
    }

}
